mirgecom/viscous.py

```diff
@@ -0,0 +1,419 @@
+        Fluid state
+    grad_cv: :class:`~mirgecom.fluid.ConservedVars`
+        Gradient of the fluid state
+    grad_t: numpy.ndarray
+        Gradient of the fluid temperature
+
+    Returns
+    -------
+    :class:`~mirgecom.fluid.ConservedVars` or float
+        The viscous transport flux vector if viscous transport properties
+        are provided, scalar zero otherwise.
+    """
+    transport = eos.transport_model()
+    if transport is None:
+        return 0
+
+    dim = cv.dim
+    viscous_mass_flux = 0 * cv.momentum
+
+    j = diffusive_flux(discr, eos, cv, grad_cv)
+    heat_flux_diffusive = diffusive_heat_flux(discr, eos, cv, j)
+
+    tau = viscous_stress_tensor(discr, eos, cv, grad_cv)
+    viscous_energy_flux = (
+        np.dot(tau, cv.velocity)
+        - conductive_heat_flux(discr, eos, cv, grad_t)
+        - heat_flux_diffusive
+    )
+
+    return make_conserved(dim,
+            mass=viscous_mass_flux,
+            energy=viscous_energy_flux,
+            momentum=tau, species_mass=-j)
+
+
+def viscous_facial_flux(discr, eos, cv_tpair, grad_cv_tpair, grad_t_tpair,
+                        local=False):
+    """Return the viscous flux across a face given the solution on both sides.
+
+    Parameters
+    ----------
+    discr: :class:`grudge.eager.EagerDGDiscretization`
+        The discretization to use
+    eos: :class:`~mirgecom.eos.GasEOS`
+        A gas equation of state
+    cv_tpair: :class:`grudge.trace_pair.TracePair`
+        Trace pair of :class:`~mirgecom.fluid.ConservedVars` with the fluid solution
+        on the faces
+    grad_cv_tpair: :class:`grudge.trace_pair.TracePair`
+        Trace pair of :class:`~mirgecom.fluid.ConservedVars` with the gradient of the
+        fluid solution on the faces
+    grad_t_tpair: :class:`grudge.trace_pair.TracePair`
+        Trace pair of temperature gradient on the faces.
+    local: bool
+        Indicates whether to skip projection of fluxes to "all_faces" or not. If
+        set to *False* (the default), the returned fluxes are projected to
+        "all_faces".  If set to *True*, the returned fluxes are not projected to
+        "all_faces"; remaining instead on the boundary restriction.
+
+    Returns
+    -------
+    :class:`~mirgecom.fluid.ConservedVars`
+        The viscous transport flux in the face-normal direction on "all_faces" or
+        local to the sub-discretization depending on *local* input parameter
+    """
+    actx = cv_tpair.int.array_context
+    normal = thaw(actx, discr.normal(cv_tpair.dd))
+
+    # todo: user-supplied flux routine
+    f_int = viscous_flux(discr, eos, cv_tpair.int, grad_cv_tpair.int,
+                         grad_t_tpair.int)
+    f_ext = viscous_flux(discr, eos, cv_tpair.ext, grad_cv_tpair.ext,
+                         grad_t_tpair.ext)
+    # This bit hard-codes BR1 central flux, use mirgecom.flux.central?
+    f_avg = 0.5*(f_int + f_ext)
+    flux_weak = make_conserved(cv_tpair.int.dim, q=(f_avg.join() @ normal))
```

Comment:
Shouldn't we use the `gradient_flux_central` routine here?

Reply:
Of course not.  The operand here is a vector so we need to use `divergence_flux_central`.

Reply:
(8a6bb2d)